File: DataStructures/Stack/Stack.ts
type Node<T> = {
    value: T;
    prev?: Node<T>;
}

export default class Stack<T> {
    public length: number = 0;
    public head?: Node<T> = undefined;

    push(item: T): void {
        this.length++;
        const node: Node<T> = { value: item };

        if (!this.head) {
            this.head = node;
            return;
        }

        node.prev = this.head;
        this.head = node;
    }

    pop(): T | undefined {
        if (!this.head)
            return undefined;

        this.length--;
        
        const prevHead = this.head;
        this.head = this.head.prev;
        prevHead.prev = undefined;

        if (this.length === 0)
            this.head = undefined;

        return prevHead.value;
    }

    peek = (): T | undefined => this.head?.value;
}

File: CSharp/DataStructures/Stack/Stack.cs
namespace CSharp.DataStructures.Stack;

public class Node<T>(T value)
{
    public T Value = value;
    public Node<T>? Prev;
}

public class Stack<T>
{
    public int Size;
    public Node<T>? Head;

    public void Push(T value)
    {
        Size++;

        var node = new Node<T>(value);

        if (Head != null)
            node.Prev = Head;

        Head = node;
    }

    public T Pop()
    {
        var head = PeekHead();

        Size--;

        Head = head.Prev;

        if (Size == 0)
            Head = null;

        return head.Value;
    }

    public T Peek()
    {
        return PeekHead().Value;
    }

    private Node<T> PeekHead()
    {
        if (Head == null)
            throw new InvalidOperationException();

        return Head;
    }
}
